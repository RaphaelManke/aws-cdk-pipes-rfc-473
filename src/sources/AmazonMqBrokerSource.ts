import { CfnBroker } from 'aws-cdk-lib/aws-amazonmq';
import { IRole, Policy, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { CfnPipe } from 'aws-cdk-lib/aws-pipes';
import { ISecret } from 'aws-cdk-lib/aws-secretsmanager';
import { IPipeSource, IPipeSourceCommonParameters } from '../PipeSource';

export interface IMQBrokerAccessCredentials {
  basicAuth: ISecret;
}

export interface IAmazonMqBrokerSourceProps extends IPipeSourceCommonParameters {
  /**
   * `CfnPipe.PipeSourceActiveMQBrokerParametersProperty.Credentials`
   *
   * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipesourceactivemqbrokerparameters.html#cfn-pipes-pipe-pipesourceactivemqbrokerparameters-credentials
   */
  //  The ARN of the Secrets Manager secret.


  readonly credentials: IMQBrokerAccessCredentials;
  /**
   * `CfnPipe.PipeSourceActiveMQBrokerParametersProperty.QueueName`
   *
   * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipesourceactivemqbrokerparameters.html#cfn-pipes-pipe-pipesourceactivemqbrokerparameters-queuename
   */
  readonly queueName: string;
};

export abstract class AmazonMqBrokerSource implements IPipeSource {
  private readonly broker: CfnBroker;
  private readonly basicAuthSecret: ISecret;
  sourceArn: string;
  sourceParameters?: CfnPipe.PipeSourceParametersProperty;

  constructor(broker: CfnBroker, props: IAmazonMqBrokerSourceProps) {

    this.basicAuthSecret = props.credentials.basicAuth;

    this.broker = broker;
    this.sourceArn = broker.attrArn;


  }

  private brokerGrantRead(grantee: IRole): void {
    grantee.attachInlinePolicy(new Policy(grantee, 'PipeSourcePolicy', {
      statements: [
        new PolicyStatement({
          actions: ['mq:Connect'],
          resources: [this.broker.attrArn],
        }),
      ],
    }));
  }

  grantRead(grantee: IRole): void {
    this.basicAuthSecret.grantRead(grantee);
    this.brokerGrantRead(grantee);

  }
}
