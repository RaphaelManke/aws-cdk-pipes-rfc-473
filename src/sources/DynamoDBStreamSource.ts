import { ITable } from 'aws-cdk-lib/aws-dynamodb';
import { IRole } from 'aws-cdk-lib/aws-iam';
import { CfnPipe } from 'aws-cdk-lib/aws-pipes';
import { PipeSource } from '../PipeSource';
import { IPipeSourceFilter } from '../PipeSourceFilter';


export interface IDynamoDBStreamSourceProps {
  filter?: IPipeSourceFilter;
  config?: CfnPipe.PipeSourceDynamoDBStreamParametersProperty;
}

export class DynamoDBStreamSource extends PipeSource {
  private table: ITable;

  constructor(table: ITable, props?: IDynamoDBStreamSourceProps) {
    super(table.tableArn, { dynamoDbStreamParameters: props?.config, filterCriteria: props?.filter });
    this.table = table;
  }

  public grantRead(grantee: IRole): void {
    this.table.grantStreamRead(grantee);
  }
}