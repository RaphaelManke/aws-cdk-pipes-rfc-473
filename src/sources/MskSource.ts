import { IResolvable, isResolvableObject, Token } from 'aws-cdk-lib';
import { IRole, Policy, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { CfnCluster, CfnServerlessCluster } from 'aws-cdk-lib/aws-msk';
import { CfnPipe } from 'aws-cdk-lib/aws-pipes';
import { ISecret } from 'aws-cdk-lib/aws-secretsmanager';
import { PipeSourceStartingPosition } from './PipeSourceStartingPosition';
import { IPipeSource, IPipeSourceCommonParameters } from '../PipeSource';


export interface IMskSourceCredentials {
  readonly clientCertificateTlsAuth?: ISecret;
  readonly saslScram512Auth?: ISecret;
}

export interface IMskSourceProps extends IPipeSourceCommonParameters {
  topicName: string;
  consumerGroupId?: string;
  startingPosition?: PipeSourceStartingPosition;
  credentials?: IMskSourceCredentials;
}


export class MskSource implements IPipeSource {
  private readonly cluster: CfnCluster | CfnServerlessCluster;
  private readonly sourceConfigSecrets: ISecret[] = [];
  sourceArn: string;
  sourceParameters?: IResolvable | CfnPipe.PipeSourceParametersProperty | undefined;

  constructor(cluster: CfnCluster | CfnServerlessCluster, props: IMskSourceProps) {
    this.cluster = cluster;
    this.sourceArn = cluster.attrArn;

    const credentials = props.credentials ? {
      clientCertificateTlsAuth: props.credentials.clientCertificateTlsAuth?.secretArn,
      saslScram512Auth: props.credentials.saslScram512Auth?.secretArn,
    } : undefined;

    if (props.credentials?.clientCertificateTlsAuth) {
      this.sourceConfigSecrets.push(props.credentials.clientCertificateTlsAuth);
    }

    if (props.credentials?.saslScram512Auth) {
      this.sourceConfigSecrets.push(props.credentials.saslScram512Auth);
    }


    this.sourceParameters = {
      managedStreamingKafkaParameters: {
        topicName: props.topicName,
        batchSize: props.batchSize,
        maximumBatchingWindowInSeconds: props?.maximumBatchingWindowInSeconds,
        credentials: credentials,
        consumerGroupId: props?.consumerGroupId,
        startingPosition: props?.startingPosition,

      },
    };
  }


  grantRead(grantee: IRole): void {
    const policies = createReadPolicyStatements(this.cluster);
    const policy = new Policy(grantee, 'PipeSourcePolicy', {
      statements: policies,
    });
    grantee.attachInlinePolicy(policy);

    this.sourceConfigSecrets.forEach((secret) =>
      secret.grantRead(grantee),
    );
  }

}


const createReadPolicyStatements = (cluster: CfnCluster | CfnServerlessCluster) : PolicyStatement[] => {

  const statements: PolicyStatement[] = [];

  statements.push(new PolicyStatement({
    sid: 'AllowReadKafkaCluster',
    actions: ['kafka:DescribeCluster',
      'kafka:DescribeClusterV2',
      'kafka:GetBootstrapBrokers'],
    resources: [cluster.attrArn],
  }));

  statements.push(new PolicyStatement({
    sid: 'AllowReadKafkaNetwork',
    actions: [
      'ec2:DescribeNetworkInterfaces',
      'ec2:DescribeSubnets',
      'ec2:DescribeSecurityGroups',
      'ec2:DescribeVpcs',
    ],
    resources: ['*'],
  }));


  let subnetIds: string[] = [];
  if (cluster instanceof CfnCluster) {
    // TODO: make a proper Resolvable check
    if ( cluster.brokerNodeGroupInfo instanceof Token || !isResolvableObject(cluster.brokerNodeGroupInfo)) {
      throw new Error('Unable to resolve brokerNodeGroupInfo');
    }

    const brokerNodeGroupInfo: CfnCluster.BrokerNodeGroupInfoProperty =
    cluster.brokerNodeGroupInfo as unknown as CfnCluster.BrokerNodeGroupInfoProperty ;
    subnetIds = brokerNodeGroupInfo.clientSubnets;

  } else {
    if ( cluster.vpcConfigs instanceof Token || !isResolvableObject(cluster.vpcConfigs)) {
      throw new Error('Unable to resolve brokerNodeGroupInfo');
    }

    const vpcConfig: CfnServerlessCluster.VpcConfigProperty =
      cluster.vpcConfigs as unknown as CfnServerlessCluster.VpcConfigProperty ;
    subnetIds = vpcConfig.subnetIds;
  }

  statements.push(new PolicyStatement({
    sid: 'AllowCreateNetworkInterfaces',
    actions: [
      'ec2:CreateNetworkInterface',
      'ec2:DeleteNetworkInterface',
    ],
    resources: ['*'],
    conditions: {
      StringEqualsIfExists: {
        'ec2:SubnetID': subnetIds,
      },
    },
  }));


  return statements;
};