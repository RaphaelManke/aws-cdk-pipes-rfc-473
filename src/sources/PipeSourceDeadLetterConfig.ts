import { Queue } from 'aws-cdk-lib/aws-sqs';


export interface PipeSourceDeadLetterConfig {
  /**
     * The ARN of the Amazon SQS queue specified as the target for the dead-letter queue.
     * see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-deadletterconfig.html#cfn-pipes-pipe-deadletterconfig-arn
     */
  queue: Queue;
}
