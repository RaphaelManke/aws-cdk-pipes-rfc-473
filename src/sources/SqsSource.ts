import { IRole } from 'aws-cdk-lib/aws-iam';
import { CfnPipe } from 'aws-cdk-lib/aws-pipes';
import { IQueue } from 'aws-cdk-lib/aws-sqs';
import { PipeSource } from '../PipeSource';
import { IPipeFilterPattern, IPipeSourceFilter, PipeGenericFilterPattern } from '../PipeSourceFilter';


export interface ISqsSourceProps {
  filter?: IPipeSourceFilter;
  config?: CfnPipe.PipeSourceSqsQueueParametersProperty;
}

export class SqsSource extends PipeSource {
  private queue: IQueue;

  constructor(queue: IQueue, props?:ISqsSourceProps) {
    super(queue.queueArn, { sqsQueueParameters: props?.config, filterCriteria: props?.filter });
    this.queue = queue;

  }

  public grantRead(grantee: IRole): void {
    this.queue.grantConsumeMessages(grantee);
  }
}


export interface ISqsAttributes {
  approximateReceiveCount?: string;
  sentTimestamp?: string;
  sequenceNumber?: string;
  messageGroupId?: string;
  senderId?: string;
  messageDeduplicationId?: string;
  approximateFirstReceiveTimestamp?: string;
}

export interface ISqsMessagePipeFilter {
  messageId?: string;
  receiptHandle?: string;
  body?: any;
  attributes?: ISqsAttributes;
  messageAttributes?: any;
  md5OfBody?: string;
}

export class PipeSqsFilterPattern extends PipeGenericFilterPattern {
  static fromSqsMessageAttributes(
    attributes: ISqsMessagePipeFilter,
  ): IPipeFilterPattern {
    const sqsProps = {
      ApproximateReceiveCount: attributes.attributes?.approximateReceiveCount,
      SentTimestamp: attributes.attributes?.sentTimestamp,
      SequenceNumber: attributes.attributes?.sequenceNumber,
      MessageGroupId: attributes.attributes?.messageGroupId,
      SenderId: attributes.attributes?.senderId,
      MessageDeduplicationId: attributes.attributes?.messageDeduplicationId,
      ApproximateFirstReceiveTimestamp:
        attributes.attributes?.approximateFirstReceiveTimestamp,
    };

    return {
      pattern: JSON.stringify({
        ...attributes,
        attributes: attributes.attributes ? sqsProps : undefined,
      }),
    };
  }
}