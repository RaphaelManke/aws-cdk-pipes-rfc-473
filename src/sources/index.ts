export * from './ActiveMqBrokerSource';
export * from './AmazonMqBrokerSource';
export * from './DynamoDBStreamSource';
export * from './KinesisStreamSource';
export * from './MskSource';
export * from './PipeSourceDeadLetterConfig';
export * from './PipeSourceOnPartialBatchItemFailure';
export * from './PipeSourceStartingPosition';
export * from './RabbitMqBrokerSource';
export * from './SelfMangedKafkaSource';
export * from './SqsSource';