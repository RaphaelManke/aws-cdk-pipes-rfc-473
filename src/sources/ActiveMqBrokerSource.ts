import { CfnBroker } from 'aws-cdk-lib/aws-amazonmq';
import { AmazonMqBrokerSource, IAmazonMqBrokerSourceProps } from './AmazonMqBrokerSource';

export class ActiveMqBrokerSource extends AmazonMqBrokerSource {
  constructor(broker: CfnBroker, props: IAmazonMqBrokerSourceProps) {
    super(broker, props);
    this.sourceParameters = {
      activeMqBrokerParameters: {
        batchSize: props.batchSize,
        maximumBatchingWindowInSeconds: props.maximumBatchingWindowInSeconds,
        credentials: { basicAuth: props.credentials.basicAuth.secretArn },
        queueName: props.queueName,
      },
    };
  }
}