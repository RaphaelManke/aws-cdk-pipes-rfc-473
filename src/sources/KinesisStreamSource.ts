import { IResolvable } from 'aws-cdk-lib';
import { IRole } from 'aws-cdk-lib/aws-iam';
import { Stream } from 'aws-cdk-lib/aws-kinesis';
import { CfnPipe } from 'aws-cdk-lib/aws-pipes';
import { IQueue } from 'aws-cdk-lib/aws-sqs';
import { PipeSourceDeadLetterConfig } from './PipeSourceDeadLetterConfig';
import { PipeSourceOnPartialBatchItemFailure } from './PipeSourceOnPartialBatchItemFailure';
import { PipeSourceStartingPosition } from './PipeSourceStartingPosition';
import { IPipeSource, IPipeSourceCommonParameters } from '../PipeSource';

export interface IKinesisStreamSourceProps extends IPipeSourceCommonParameters {

  /**
      * `CfnPipe.PipeSourceKinesisStreamParametersProperty.DeadLetterConfig`
      *
      * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipesourcekinesisstreamparameters.html#cfn-pipes-pipe-pipesourcekinesisstreamparameters-deadletterconfig
      */
  readonly deadLetterConfig?: PipeSourceDeadLetterConfig;

  /**
      * `CfnPipe.PipeSourceKinesisStreamParametersProperty.MaximumRecordAgeInSeconds`
      *
      * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipesourcekinesisstreamparameters.html#cfn-pipes-pipe-pipesourcekinesisstreamparameters-maximumrecordageinseconds
      */
  readonly maximumRecordAgeInSeconds?: number;
  /**
      * `CfnPipe.PipeSourceKinesisStreamParametersProperty.MaximumRetryAttempts`
      *
      * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipesourcekinesisstreamparameters.html#cfn-pipes-pipe-pipesourcekinesisstreamparameters-maximumretryattempts
      */
  readonly maximumRetryAttempts?: number;
  /**
      * `CfnPipe.PipeSourceKinesisStreamParametersProperty.OnPartialBatchItemFailure`
      *
      * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipesourcekinesisstreamparameters.html#cfn-pipes-pipe-pipesourcekinesisstreamparameters-onpartialbatchitemfailure
      */
  readonly onPartialBatchItemFailure?: PipeSourceOnPartialBatchItemFailure;
  /**
      * `CfnPipe.PipeSourceKinesisStreamParametersProperty.ParallelizationFactor`
      *
      * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipesourcekinesisstreamparameters.html#cfn-pipes-pipe-pipesourcekinesisstreamparameters-parallelizationfactor
      */
  readonly parallelizationFactor?: number;
  /**
      * `CfnPipe.PipeSourceKinesisStreamParametersProperty.StartingPosition`
      *
      * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipesourcekinesisstreamparameters.html#cfn-pipes-pipe-pipesourcekinesisstreamparameters-startingposition
      */
  readonly startingPosition: PipeSourceStartingPosition;
  /**
      * `CfnPipe.PipeSourceKinesisStreamParametersProperty.StartingPositionTimestamp`
      *
      * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipesourcekinesisstreamparameters.html#cfn-pipes-pipe-pipesourcekinesisstreamparameters-startingpositiontimestamp
      */
  readonly startingPositionTimestamp?: string;
}


export class KinesisStreamSource implements IPipeSource {
  private readonly stream: Stream;
  private readonly dlq?: IQueue;

  sourceArn: string;
  sourceParameters?: IResolvable | CfnPipe.PipeSourceParametersProperty | undefined;

  constructor(kinesisStream: Stream, props: IKinesisStreamSourceProps) {
    if (props.startingPosition === PipeSourceStartingPosition.AT_TIMESTAMP && !props.startingPositionTimestamp) {
      throw new Error('startingPositionTimestamp must be set when startingPosition is AT_TIMESTAMP');
    }
    this.stream = kinesisStream;

    const deadLetterConfig = props.deadLetterConfig ? {
      arn: props.deadLetterConfig.queue.queueArn,
    } : undefined;

    if ( props.deadLetterConfig?.queue ) {
      this.dlq = props.deadLetterConfig?.queue;
    }

    this.sourceArn = kinesisStream.streamArn;
    this.sourceParameters = {
      kinesisStreamParameters: {
        deadLetterConfig: deadLetterConfig,
        maximumRecordAgeInSeconds: props.maximumRecordAgeInSeconds,
        maximumRetryAttempts: props.maximumRetryAttempts,
        onPartialBatchItemFailure: props.onPartialBatchItemFailure,
        parallelizationFactor: props.parallelizationFactor,
        startingPositionTimestamp: props.startingPositionTimestamp,
        batchSize: props.batchSize,
        maximumBatchingWindowInSeconds: props.maximumBatchingWindowInSeconds,
        startingPosition: props.startingPosition,
      },
    };
  }
  grantRead(grantee: IRole): void {
    this.stream.grantRead(grantee);
    if (this.dlq) {
      this.dlq.grantSendMessages(grantee);
    }
  }
}