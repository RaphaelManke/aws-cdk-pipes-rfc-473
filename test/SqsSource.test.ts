import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { CfnPipe } from 'aws-cdk-lib/aws-pipes';
import { Queue } from 'aws-cdk-lib/aws-sqs';
import { SqsSource } from '../src/sources/SqsSource';

describe('SqsSource', () => {
  it('should synth', () => {
    // ARRANGE
    const stack = new Stack(new App(), 'test-stack');
    const sourceQueue = new Queue(stack, 'test-source-queue', {});

    // ACT
    const source = new SqsSource(sourceQueue);
    new CfnPipe(stack, 'test-pipe', {
      roleArn: 'test-role-arn',
      source: source.sourceArn,
      target: 'test-target-arn',
      sourceParameters: source.sourceParameters,
    },
    );

    // ASSERT
    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();

  });

  it('should accept filter', () => {
    // ARRANGE
    const stack = new Stack(new App(), 'test-stack');
    const sourceQueue = new Queue(stack, 'test-source-queue', {});

    // ACT
    const source = new SqsSource(sourceQueue, {
      filterCriteria: {
        filters: [
          {
            pattern: JSON.stringify({
              foo: 'bar',
            }),
          },
        ],
      },
    });

    new CfnPipe(stack, 'test-pipe', {
      roleArn: 'test-role-arn',
      source: source.sourceArn,
      target: 'test-target-arn',
      sourceParameters: source.sourceParameters,
    },
    );

    // ASSERT
    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();


  });

  it('should accept config', () => {
    // ARRANGE
    const stack = new Stack(new App(), 'test-stack');
    const sourceQueue = new Queue(stack, 'test-source-queue', {});

    // ACT
    const source = new SqsSource(sourceQueue, {
      batchSize: 10,
      maximumBatchingWindowInSeconds: 10,
    });
    new CfnPipe(stack, 'test-pipe', {
      roleArn: 'test-role-arn',
      source: source.sourceArn,
      target: 'test-target-arn',
      sourceParameters: source.sourceParameters,
    },
    );

    // ASSERT
    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();
  });
});