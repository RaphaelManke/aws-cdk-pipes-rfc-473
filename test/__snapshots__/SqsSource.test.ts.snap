// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SqsSource should accept config 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "testpipe": Object {
      "Properties": Object {
        "RoleArn": "test-role-arn",
        "Source": Object {
          "Fn::GetAtt": Array [
            "testsourcequeueDAC45857",
            "Arn",
          ],
        },
        "SourceParameters": Object {
          "SqsQueueParameters": Object {
            "BatchSize": 10,
            "MaximumBatchingWindowInSeconds": 10,
          },
        },
        "Target": "test-target-arn",
      },
      "Type": "AWS::Pipes::Pipe",
    },
    "testsourcequeueDAC45857": Object {
      "DeletionPolicy": "Delete",
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`SqsSource should accept filter 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "testpipe": Object {
      "Properties": Object {
        "RoleArn": "test-role-arn",
        "Source": Object {
          "Fn::GetAtt": Array [
            "testsourcequeueDAC45857",
            "Arn",
          ],
        },
        "SourceParameters": Object {
          "FilterCriteria": Object {
            "Filters": Array [
              Object {
                "Pattern": "{\\"foo\\":\\"bar\\"}",
              },
            ],
          },
        },
        "Target": "test-target-arn",
      },
      "Type": "AWS::Pipes::Pipe",
    },
    "testsourcequeueDAC45857": Object {
      "DeletionPolicy": "Delete",
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`SqsSource should synth 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "testpipe": Object {
      "Properties": Object {
        "RoleArn": "test-role-arn",
        "Source": Object {
          "Fn::GetAtt": Array [
            "testsourcequeueDAC45857",
            "Arn",
          ],
        },
        "SourceParameters": Object {},
        "Target": "test-target-arn",
      },
      "Type": "AWS::Pipes::Pipe",
    },
    "testsourcequeueDAC45857": Object {
      "DeletionPolicy": "Delete",
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
